### DISCUSSION TITLE
registerTarget capability dropped in new plugin API

### DISCUSSION THREAD

**Discussion by cdaringe** [nrwl/nx#22549] [General]
### Current Behavior

- capability to augment `nx`'s registered project model with new targets is not present in the new (v2?) plugin api

### Expected Behavior

- capability to augment `nx`'s registered project model

### GitHub Repo

n/a

### Steps to Reproduce

1. Attempt to add a new `target` to an existing model
2. Observe that no such API exists

https://nx.dev/extending-nx/recipes/project-graph-plugins recommends using `createNodes` or `createDependencies`.

Neither of these replace the excellent behavior of the prior API.

I do not want to create a project (`createNodes`) or add a dependency (`createDependencies`).

I simply want `nx`'s hydrated model, and to augment it with a new target. This is nice because `nx` now loads models from project.json and package.json.

Some top-level plugin API should allow my plugin to go nuts with the finalized nx project model.

### Nx Report

```shell
n/a
```

### Failure Logs

```shell
n/a
```

### Package Manager Version

n/a

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by FrozenPandaz** [comment]
It was intentional that we rolled this into the `createNodes` API.

The API is perhaps not as simple as the `registerTarget` API. However, the new API is more powerful and allows you to do even more than before. We've been successful in using the `createNodes` API for many plugins now.

Let's discuss what you are trying to do and how you can do it with the new API.

cc @AgentEnder

**Reply by cdaringe** [reply]
Currently, I discover targets from `package.json::scripts`, e.g. `test` and `lint`. In this project, I am not using `project.json`, etc.

I want to add a `codegen` script to every project via plugin.  I do not see a means to replicate this capability.

**Reply by AgentEnder** [reply]
A plugin that adds a target to all package.json projects would look something like this:

```typescript
export const createNodes: CreateNodes = ['**/package.json', (packageJsonPath, opts, ctx) => {
  const packageJson = readFileSync(join(ctx.workspaceRoot, packageJsonPath));
  const projectRoot = dirname(packageJsonPath);
  const myTargetConfiguration = someMethodToGetTarget(packageJson);

  return {
    projects: {
      [projectRoot]: {
        targets: {
          myTarget: myTargetConfiguration
        }
      }
    }
  }
}];
```

**Reply by AgentEnder** [reply]
Additionally, any v1 inference plugin can be converted to a v2 plugin without losing any functionality (see https://github.com/nrwl/nx/blob/18.2.0/packages/nx/src/utils/nx-plugin.ts#L308-L333)
